pkg/scalers/predictkube_scaler.go

```diff
@@ -0,0 +1,463 @@
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	str2duration "github.com/xhit/go-str2duration/v2"
+	"google.golang.org/grpc"
+	health "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+	"k8s.io/api/autoscaling/v2beta2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	predictKubeMetricType = "External"
+)
+
+var (
+	mlEngineHost = "predictkube-dev.dysnix.org"
+	mlEnginePort = 8080
+
+	defaultStep = time.Minute * 5
+)
+
+type PredictKubeScaler struct {
+	metadata         *predictKubeMetadata
+	prometheusClient api.Client
+	grpcConn         *grpc.ClientConn
+	grpcClient       pb.MlEngineServiceClient
+	healthClient     health.HealthClient
+	api              v1.API
+	transport        http_transport.FastHttpTransport
+}
+
+type predictKubeMetadata struct {
+	predictHorizon    time.Duration
+	historyTimeWindow time.Duration
+	stepDuration      time.Duration
+	apiKey            string
+	prometheusAddress string
+	metricName        string
+	query             string
+	threshold         int64
+	scalerIndex       int
+}
+
+var predictKubeLog = logf.Log.WithName("predictkube_scaler")
+
+func (pks *PredictKubeScaler) setupClientConn() error {
+	clientOpt, err := pc.SetGrpcClientOptions(&libs.GRPC{
+		Enabled:       true,
+		UseReflection: true,
+		Compression: libs.Compression{
+			Enabled: false,
+		},
+		Conn: &libs.Connection{
+			Host:            mlEngineHost,
+			Port:            uint16(mlEnginePort),
+			ReadBufferSize:  50 << 20,
+			WriteBufferSize: 50 << 20,
+			MaxMessageSize:  50 << 20,
+			Insecure:        true,
+			Timeout:         time.Second * 15,
+		},
+		Keepalive: &libs.Keepalive{
+			Time:    time.Minute * 5,
+			Timeout: time.Minute * 5,
+			EnforcementPolicy: &libs.EnforcementPolicy{
+				MinTime:             time.Minute * 20,
+				PermitWithoutStream: false,
+			},
+		},
+	},
+		&libs.Base{
+			Monitoring: libs.Monitoring{
+				Enabled: false,
+			},
+			Profiling: libs.Profiling{
+				Enabled: false,
+			},
+			IsDebugMode: true,
+			Single: &libs.Single{
+				Enabled: false,
+			},
+		},
+		pc.InjectPublicClientMetadataInterceptor(pks.metadata.apiKey),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	pks.grpcConn, err = grpc.Dial(fmt.Sprintf("%s:%d", mlEngineHost, mlEnginePort), clientOpt...)
+	if err != nil {
+		return err
+	}
+
+	pks.grpcClient = pb.NewMlEngineServiceClient(pks.grpcConn)
+	pks.healthClient = health.NewHealthClient(pks.grpcConn)
+
+	return err
+}
+
+// NewPredictKubeScaler creates a new PredictKube scaler
+func NewPredictKubeScaler(_ context.Context, config *ScalerConfig) (*PredictKubeScaler, error) {
+	s := &PredictKubeScaler{}
```

Comment:
is there any reason for not following the pattern that is in use in other scalers? I mean, the rest of the scalers returns a new object from the parsed data (but maybe here we can for any reason)

Reply:
We use context in `initPredictKubePrometheusConn` method as you requested below.